src/communication/v2/session.hpp

```diff
@@ -0,0 +1,331 @@
+      spdlog::error("Session: Destructor called while execution is active");
+    }
+  }
+
+  bool Start() {
+    if (execution_active_) {
+      return false;
+    }
+    execution_active_ = true;
+    timeout_timer_.async_wait(
+        boost::asio::bind_executor(strand_, std::bind(&Session::OnTimeout, this->shared_from_this())));
+
+    if (auto *socket = std::get_if<SSLSocket>(&socket_); socket) {
+      boost::asio::dispatch(this->shared_from_this()->strand_,
+                            [shared_this = this->shared_from_this()] { shared_this->DoHandshake(); });
+    } else {
+      boost::asio::dispatch(this->shared_from_this()->strand_,
+                            [shared_this = this->shared_from_this()] { shared_this->DoRead(); });
+    }
+    return true;
+  }
+
+  bool Write(const uint8_t *data, size_t len, bool have_more = false) {
+    if (!IsConnected()) {
+      return false;
+    }
+    boost::asio::dispatch(this->strand_, [shared_this = this->shared_from_this(), data, len, have_more] {
+      shared_this->DoWrite(data, len, have_more);
+    });
+    return true;
+  }
+
+  bool IsConnected() const {
+    return std::visit(
+        utils::Overloaded{
+            [this](const SSLSocket &socket) { return execution_active_ && socket.lowest_layer().is_open(); },
+            [this](const TCPSocket &socket) { return execution_active_ && socket.is_open(); }},
+        socket_);
+  }
+
+ private:
+  explicit Session(tcp::socket &&socket, TSessionData *data, ServerContext &server_context, tcp::endpoint endpoint,
+                   int inactivity_timeout_sec, std::string_view service_name)
+      : socket_(CreateWebSocket(std::move(socket), server_context)),
+        strand_{boost::asio::make_strand(GetExecutor())},
+        output_stream_([this](const uint8_t *data, size_t len, bool have_more) { return Write(data, len, have_more); }),
+        session_(data, endpoint, input_buffer_.read_end(), &output_stream_),
+        endpoint_{endpoint},
+        remote_endpoint_{GetRemoteEndpoint()},
+        service_name_{service_name},
+        timeout_seconds_(inactivity_timeout_sec),
+        timeout_timer_(GetExecutor()) {
+    std::visit(
+        utils::Overloaded{[](SSLSocket &socket) {
+                            socket.lowest_layer().set_option(boost::asio::ip::tcp::no_delay(true));  // enable PSH
+                            socket.lowest_layer().set_option(
+                                boost::asio::socket_base::keep_alive(true));  // enable SO_KEEPALIVE
+                            socket.lowest_layer().non_blocking(false);
+                          },
+                          [](TCPSocket &socket) {
+                            socket.set_option(boost::asio::ip::tcp::no_delay(true));        // enable PSH
+                            socket.set_option(boost::asio::socket_base::keep_alive(true));  // enable SO_KEEPALIVE
+                            socket.non_blocking(false);
+                          }},
+        socket_);
```

Comment:
I've tried doing something like that but I've encountered issues, so let me explain:
1. `lowest_layer()` does not return TCP socket but the lowest layer type, so the function LowestLayer does not have the same return type
2. ok then I've tried letting the compiler deduce which return type it is
```c++
  auto &LowestLayer() {
    return std::visit(utils::Overloaded{[](TCPSocket &socket) { return socket; },
                                        [](SSLSocket &socket) { return socket.lowest_layer(); }},
                      socket_);
  }
```
and here I get the error that I am trying to copy socket, which is understandable since that can happen from the Overload lambda call,
3. Then I've tried:
```c++
  auto &LowestLayer() {
    return std::visit(utils::Overloaded{[](TCPSocket &socket) -> auto& { return socket; },
                                        [](SSLSocket &socket) -> auto& { return socket.lowest_layer(); }},
                      socket_);
  }
```
I've also tried with `TCPSocket&` and `SSLSocket::lowest_layer_type&` as return types but I get  `non-void constexpr function 'visit...` errors. Not sure if I am missing something?

Reply:
Actually it is even easier! `lowest_layer` is defined for both `TCPSocket` and `SSLSocket`, so it can be written in this way:
```
    std::visit(utils::Overloaded{[](auto &socket) {
                 auto &lowest_layer = socket.lowest_layer();
                 lowest_layer.set_option(boost::asio::ip::tcp::no_delay(true));        // enable PSH
                 lowest_layer.set_option(boost::asio::socket_base::keep_alive(true));  // enable SO_KEEPALIVE
                 lowest_layer.non_blocking(false);
               }},
               socket_);
```

Proof:
`boost::asio::ip::tcp::socket` is defined as `basic_stream_socket<tcp>`, and `basic_stream_socket` is derived from `basic_socket`, so `basic_socket::lowest_layer` can be used for `TCPSocket`.
```
  /// Get a reference to the lowest layer.
  /**
   * This function returns a reference to the lowest layer in a stack of
   * layers. Since a basic_socket cannot contain any further layers, it simply
   * returns a reference to itself.
   *
   * @return A reference to the lowest layer in the stack of layers. Ownership
   * is not transferred to the caller.
   */
  lowest_layer_type& lowest_layer()
  {
    return *this;
  }
```
For `SSLSocket` which is `boost::asio::ssl::stream<TCPSocket>`, the `lowest_layer` is defined in the following way:
```
  /// Get a reference to the lowest layer.
  /**
   * This function returns a reference to the lowest layer in a stack of
   * stream layers.
   *
   * @return A reference to the lowest layer in the stack of stream layers.
   * Ownership is not transferred to the caller.
   */
  lowest_layer_type& lowest_layer()
  {
    return next_layer_.lowest_layer();
  }
```

I tried it, and it worked for me, so I think it should work in general. What do you think?

Reply:
Aha, you are right, I like this approach much better!